server/app/controllers/applicant/UpsellController.java

```diff
@@ -165,4 +177,64 @@ public CompletionStage<Result> considerRegister(
               throw new RuntimeException(ex);
             });
   }
+
+  /** Download a PDF file of the application to the program. */
+  @Secure
+  public CompletionStage<Result> download(
+      Http.Request request, long programId, long applicationId, long applicantId)
+      throws ProgramNotFoundException {
+    CompletableFuture<ProgramDefinition> program =
+        programService.getProgramDefinitionAsync(programId).toCompletableFuture();
+    Optional<CiviFormProfile> profileMaybe = profileUtils.currentUserProfile(request);
+    CiviFormProfile profile =
+        profileMaybe.orElseThrow(
+            () -> new NoSuchElementException("User authorized as applicant but no profile found"));
+
+    CompletableFuture<ApplicantPersonalInfo> applicantPersonalInfo =
+        applicantService.getPersonalInfo(applicantId).toCompletableFuture();
+
+    CompletableFuture<Account> account =
+        applicantPersonalInfo
+            .thenComposeAsync(
+                v -> checkApplicantAuthorization(request, applicantId), httpContext.current())
```

Comment:
**suggestion (blocking):** there is still a security vulnerability here because it's using the user-provided `applicantId`. Instead use the applicant ID from the `account` resolve via their `profile` i.e. their session cookie.

Reply:
We cannot use this approach as it won't work in TI path. When a TI is applying on behalf of a client, the Profile is still TI, but the applicant Id belongs to the Client.

Reply:
Gotcha, in the case the `applicantId` provided in the URL needs to be validated as either the current user's applicant ID or as associated with an account that is managed by the current user's TI org.

Reply:
I think the checkApplicantAuthorization does it. @zarifzm - please confirm

Reply:
It does. https://github.com/civiform/civiform/blob/main/server/app/auth/CiviFormProfile.java#L203